BIP Draft: Testnet 5 (#2196)
bitcoin/bips: bip-0095.md

transactions that other miners would not relay, and to keep the chain moving if a large source of hash
power were to leave the network. Shortly after Testnet 4's introduction, the exception has been
systematically and sustainably exploited, which prevented the exception from achieving the intended
goals. While block storms were prevented, the network suffers from constant re-orgs of small
numbers of blocks due to multiple difficulty-exception blocks competing for the tip. This led
to discussion about changing Testnet 4 to mitigate this issue (see [Bitcointalk][bitcointalk-thread]
for analysis and discussion).

In Testnet 5 there is no exception to the PoW rules. This appears to be the logical conclusion,
since any such exception could be exploited by a motivated attacker. This
ensures the network’s behavior matches mainnet as closely as possible.

BIP 54 is already enforced on signet through [Bitcoin Inquisition][signet-bip54] as of this BIP's creation.
However, signet does not allow miners to test that their software reliably follows the rules of
BIP 54. Testnet 5 provides a testing environment for this.

## Specification

> Testnet 5 follows the same consensus rules as mainnet with the following two changes.

Reviewer: I assume that this implies that segwit and taproot are also active starting from block 1, but perhaps that should be mentioned?

Reviewer: Nevermind, I see that it is more explicitly mentioned below in Network Parameters. Perhaps Network Parameters should be a subsection of Specification, though?

Author: I moved the Network Parameters section into Specification

### Message Start

The message start is defined as <code>0x46495645</code>. These four bytes spell `FIVE` when
interpreted as ASCII.

### Network Parameters

The default p2p port for Testnet 5 is `18335`.

## Backward Compatibility

Testnet 5's consensus rules are not compatible with those of Testnet 3 and Testnet 4. The
consensus rules differ in both directions: Testnet 5 enforces the BIP 54 consensus rules from
block 1 which is not the case for Testnet 3 or Testnet 4. Testnet 5 also does not apply the
difficulty exception rule from Testnet 3 or Testnet 4 requires.

Any implementation that intends to follow Testnet 5 must add the new network parameters and
additionally enforce the BIP 54 rules while not permitting any of the difficulty exception rules.

## Reference Implementation

Pull request at ?

## References

[block-storms]: https://blog.lopp.net/the-block-storms-of-bitcoins-testnet/
[bitcointalk-thread]: https://bitcointalk.org/index.php?topic=5569103.0
[signet-bip54]: https://delvingbitcoin.org/t/bitcoin-inqusition-29-2/2236
[BIP30]: bip-0030.mediawiki
[BIP54]: bip-0054.md

> [BIP94]: bip-0094.mediawiki

Reviewer: Since these references are all just link definitions, the References section currently appears empty in the rendered document.

Author: Oh, I added a visible list here now as well.

## Copyright

This document is licensed under the Creative Commons CC0 1.0 Universal license.

## Changelog

* __0.1.0__ (2026-06-02):

>     * Initial draft

Reviewer: Nit: Changelog is only necessary if there are changes to a BIP after it has been advanced to Complete, but it is of course allowed to have one from the get-go.

However, if you do have a Changelog, please include the corresponding Version header in the preamble.

Author: I like to have a changelog now which lists the changes since the BIP PR was opened because people may read the version that was referenced in the mailing list post first and miss some changes that were made between that and the latest version here. So I added the version preamble.

```
  BIP: ?
  Layer: Applications
  Title: Testnet 5
  Authors: Pol Espinasa <polespinasa@protonmail.com>
           Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: CC0-1.0
  Discussion: 2026-06-02: https://groups.google.com/g/bitcoindev/c/kGUMTxOvdJA

>   Version: 0.1.0

Reviewer: I think the following two additional headers would make sense:

```suggestion
  Version: 0.1.0
  Requires: 54
  Replaces: 94
```

Author: Done

to use for testing. Additionally, Testnet 5 enforces the consensus rules specified in BIP 54 from block 1.

## Motivation

Testnet 4 included mitigations for an issue known as the [block storm attack][block-storms] which could render the
whole network unusable. This led to a depletion of block subsidies, which made it hard to acquire
coins for testing. However, Testnet 4 still retained a modified version of the difficulty exception rule
with the aim of allowing CPU users a limited path to acquire coins for testing, to mine non-standard
transactions that other miners would not relay, and to keep the chain moving if a large source of hash
power were to leave the network. Shortly after Testnet 4's introduction, the exception has been
systematically and persistently exploited, which prevented the exception from achieving the intended
goals. While block storms were prevented, the network suffers from constant re-orgs of small
numbers of blocks due to multiple difficulty-exception blocks competing for the tip. This led
to discussion about changing Testnet 4 to mitigate this issue (see [Bitcointalk][bitcointalk-thread]
for analysis and discussion).

> In Testnet 5 there is no exception to the PoW rules. This appears to be the logical conclusion,

Reviewer: I just had a random thought that I was wondering whether it was considered:

What would happen if blocks mined with the difficulty exception would be forbidden from collecting the subsidy (i.e., could only collect fees)? That would permit users to get blocks if the difficulty had run up, or to mine non-standard transactions, but would remove the incentive to mine blocks for collecting the subsidy. It would not prevent people from creating low-difficulty blocks to deny others from getting them and it might be more complicated than just dropping the exception, though.

Reviewer 2: Yet another band-aid fix, I fear.  Way better off fresh and "right" from the start.

There are other ways to earn coins besides Block Rewards, using a difficulty exploit.  Perhaps even more disruptive.

I also agree with the more complicated.

Reviewer 3: > What would happen if blocks mined with the difficulty exception would be forbidden from collecting the subsidy (i.e., could only collect fees)?

Then, coins will be burned, and ASIC miners will have an incentive to reorg them. To prevent that, additional coins could be simply timelocked to some future block number, then stronger miners won't have an incentive to reorg them, and will simply collect them later instead.

Anyway, the idea to decrease the coinbase reward, proportionally to the Proof of Work in the block header, is not new. But it makes things more complicated, and there are other ways, like pay to Proof of Work scripts, which could be used to achieve the same things, without messing up with consensus rules.

Also, the scenario, where the chain would halt, because of not enough miners, could happen on the mainnet as well, so it would be good, if test network would behave in a similar way, to let someone try to deal with that problem, and prepare a solution for mainnet as well.

Author: It's an interesting idea but I agree that it would be adding more complexity while this proposal is partly motivated by removing complexity. And I also think that while it may remove part of the incentive in the form of the subsidy, since this is basically free to exploit, there will still be people trying to do this just for the fees or to mess with the network for fun.

Testnet 5 follows the same consensus rules as mainnet with the following two changes.

### BIP 54 activation

The rules specified in [BIP 54 version 1.0.0][BIP54] are active on Testnet 5 from block 1.

#### Problem Statement

BIP 54 proposes new consensus rules in order to fix several potential attack vectors. Namely 
it prevents the timewarp attack, reduces the worst-case block validation time, mitigates Merkle
tree weaknesses, and avoids duplicate transactions without [bip-0030][BIP30] validation.

#### Rule Specification

See specification in [bip-0054][BIP54].

### Minimum Difficulty

The proof-of-work limit (the maximum target) is `0x1a0fffff` in compact encoding (nBits),
corresponding to a minimum difficulty of approximately 1,000,000.

### Consensus Rules

All consensus rules active on mainnet as of May 2026 are enforced from block 1, the most
recent of these being the Taproot softfork. Additionally, the rules of BIP 54 are enforced
from block 1, as specified above.

### Genesis Block

TODO: Mine the block. The values below are placeholders inherited from Testnet 4. They are
the genesis block's header fields together with the `Message` and `Pubkey` of its coinbase

> transaction. Notes for the miner:

Reviewer: I took a look at BIP94 in regard to the meaning of these two fields. It seems to me that the `pubkey` field is also only implicitly explained there. Looking at the construction, it seems to me that the `pubkey` field takes the place of the previous block hash in the block header, is that right? I was unable to find out where the `message` field goes. I thought it would appear in the coinbase field of the coinbase transaction, but I could not find the message string either forward or backward in the block hex.
I would recommend that it be explicitly explained where the `pubkey` and `message` field appear in the construction.

Reviewer: Oooh. Does `pubkey` go into the P2PK output script? Still confused about the message, though.

Reviewer 3: I guess it could be just some x-value pubkey, equal to the mainnet block hash. For example: https://mempool.space/address/0200000000000000000001ae18e25b101d485505283d8d8e67771e1d865acbfece (and then, getting the private key for the Genesis Block would be impossible, if the secp256k1 point would be invalid, or would be as hard, as breaking secp256k1, if it would turn out to be a valid point, like 04 00000000000000000001AE18E25B101D485505283D8D8E67771E1D865ACBFECE 0E007A49956DB6EF8251FADDB106C649250FD5E5EEE9E875C88F62EACF3473D4 is in this case).

Author: @murchandamus Right, see the footnote change I made, I hope that makes it clear enough. I also added details to explain the message. Since a future Testnet creator wouldn't need to use this mechanism of commitment at all anymore, I think it's fine to not do another footnote on it and remove it when the genesis block is added, but I can be convinced otherwise if people think it's interesting to keep the historical context around.

Reviewer: Thanks, looks good!

```
  BIP: ?
  Layer: Applications
  Title: Testnet 5
  Authors: Pol Espinasa <polespinasa@protonmail.com>
           Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: CC0-1.0
  Discussion: 2026-06-02: https://groups.google.com/g/bitcoindev/c/kGUMTxOvdJA
  Version: 0.1.0
  Requires: 54
  Replaces: 94
```

## Abstract

Testnet 5 is a new test network intended to replace [Testnet 4][BIP94]. Testnet 5 removes the difficulty
exception defined in Testnet 4. Sustained exploitation of this exception has made the network difficult
to use for testing. Additionally, Testnet 5 raises the minimum difficulty and enforces the consensus
rules specified in BIP 54 from block 1.

## Motivation

Testnet 4 included mitigations for an issue known as the [block storm attack][block-storms] which could render the

> whole network unusable. Block storm attacks led to a depletion of block subsidies, which made it hard to acquire

Reviewer 2: Testnet 3 never became unusable for users or especially for the "whole network"... on the contrary, confirmations came in much faster, and made using certain services easier. My testnet service did better during blockstorms for users (100 confirmations didn't take days).

Some services that had preexisting problems (expected in testing) had problems, but many did not. The ones who had issues have more Twitter followers.

What the blockstorm did was drain the block reward from miners, rendering mining basically useless. I agree with this.

Author: > Testnet 3 never became unusable for users or especially for the "whole network"

If someone with a capable machine just mines min-difficult blocks with maximum speed deactivates the mempool, the network is unusable because nobody would get their transactions in other than mining a real difficulty block. What user facing wallet would not have a problem with that?

Reviewer 2: Testnet3 didn't face min block miners purposely mining empty blocks.

Testnet3 was attacked by Lopp doing blockstorms and they included transactions.

I'm just stating the facts of what did happen.  I'm not sure what could have happened.

I'd be totally happy to provide historical blockchain evidence if you'd like.

Also, I believe the "capable machine" causing the Testnet4 issues is or was a BitAxe.  It's more about how the machine is used than about its ability when gamesmanship is involved.

Author: So then why write this comment at all? The text already says that this could happen, not that it did happen.

Reviewer 2: I wrote this comment so future contributors and interested individuals have historical facts to work with, rather than design proposals engineered around an imaginary situation that didn't actually happen.

Author: Yes, Lopp was probably still being nice when he demonstrated block storms, doesn't mean that this need to be misinterpreted as block storms being less of a problem.

Reviewer 2: I do agree it was a problem (several called it a "feature" at first). However, I think development documentation should lean toward historical facts rather than "if" and "probably."

Lopp wasn't acting out of niceness or direct financial testnet incentive, and the empty blocks on Testnet 4 aren't due to someone being mean—they are purely driven by financial incentives. The two attack vectors shouldn't be blended together. 

Anyways, no need to respond.  Thank you for not deleting these facts and taking them into consideration so that this important documentation can convey the facts of how we got here and not fiction.

coins for testing. However, Testnet 4 still retained a modified version of the difficulty exception rule.
The difficulty exception permits a block to be mined at the minimum difficulty when more than 20 minutes
have passed since the previous block. Testnet 4 kept this exception with the aim of allowing CPU users a
limited path to acquire coins for testing, to mine non-standard transactions that other miners would not
relay, and to keep the chain moving if a large source of hash power were to leave the network. Shortly

> after Testnet 4's introduction, the exception has been systematically and persistently exploited, which

Reviewer 4: ```suggestion
after Testnet 4's introduction, the exception had been systematically and persistently exploited, which
```

Author: Done

prevented the exception from achieving the intended

> goals. While block storms were prevented, the network suffers from constant re-orgs of small

Reviewer 4: s/suffers/suffered/

Author: Done

numbers of blocks due to multiple difficulty-exception blocks competing for the tip. This led
to discussion about changing Testnet 4 to mitigate this issue (see [Bitcointalk][bitcointalk-thread]
for analysis and discussion).

In Testnet 5 there is no exception to the PoW rules. This appears to be the logical conclusion,

> since any such exception could be exploited by a motivated attacker. This

Reviewer 4: Mildly ambiguous subject.

```suggestion
since any such exception could be exploited by a motivated attacker. The lack of exception
```

Author: Done

ensures the network’s behavior matches mainnet as closely as possible.

BIP 54 is already enforced on signet through [Bitcoin Inquisition][signet-bip54] as of this BIP's creation.
However, signet does not allow miners to test that their software reliably follows the rules of
BIP 54. Testnet 5 provides a testing environment for this.

## Specification

> Testnet 5 follows the same consensus rules as mainnet with the following two changes.

Reviewer 4: ```suggestion
Testnet 5 follows the same consensus rules as mainnet, except for the following two changes.
```

Author: Done

### BIP 54 activation

> The rules specified in [BIP 54 version 1.0.0][BIP54] are active on Testnet 5 from block 1.

Reviewer 5: Should the link to BIP54 be present on the first mention of the BIP (Line 22 under Abstract)?

Author: Done

> * Pubkey: <code>000000000000000000000000000000000000000000000000000000000000000000</code>
> * Time stamp: 1714777860
> * Nonce: 393743547
> * Difficulty: <code>0x1d00ffff</code>
> * Version: 1
> 
> The resulting Genesis block hash is <code>00000000da84f2bafbbc53dee25a72ae507ff4914b867c565be350b0da8bf043</code>, and the block hex is <code>0100000000000000000000000000000000000000000000000000000000000000000000004e7b2b9128fe0291db0693af2ae418b767e657cd407e80cb1434221eaea7a07a046f3566ffff001dbb0c78170101000000010000000000000000000000000000000000000000000000000000000000000000ffffffff5504ffff001d01044c4c30332f4d61792f323032342030303030303030303030303030303030303030303165626435386332343439373062336161396437383362623030313031316662653865613865393865303065ffffffff0100f2052a010000002321000000000000000000000000000000000000000000000000000000000000000000ac00000000</code>.

### Message Start

The message start (also known as the network magic) is defined as <code>0x46495645</code>. These four bytes spell `FIVE` when
interpreted as ASCII.

### P2P Port

The default p2p port for Testnet 5 is `18335`.

## Rationale

> Instead of starting a new Testnet, changing the rules of Testnet 4 was considered as well. The decision

Reviewer 4: Perhaps link to discussion in https://github.com/bitcoin/bitcoin/pull/31117 somewhere in this BIP.

```suggestion
Instead of starting a new Testnet, changing the rules of Testnet 4 [was considered](https://github.com/bitcoin/bitcoin/pull/31117) as well. The decision
```

Author: Done

> for a new network has two main reasons:

Reviewer 4: ```suggestion
for a new network had two main reasons:
```

Author: Done

   including those of BIP 54.
2. Any fix to Testnet 4 that goes beyond a band-aid would require a hard fork, which brings its own
   implementation cost: adding hard-fork support to clients, handling the p2p layer correctly, etc.
   For a test network this isn't worth taking on, especially since a band-aid would only make
   Testnet 4 diverge further from mainnet.

Raising the minimum difficulty was dismissed for Testnet 4 because it would have prevented CPU
miners from utilizing the difficulty exception. With the exception removed in Testnet 5, this
argument no longer applies. A minimum difficulty of approximately 1,000,000 strikes a balance
between dampening the quick mining of large numbers of blocks shortly after launch, before the
difficulty has adjusted to the available hash power, and keeping a low barrier so that a handful
of at-home miners or a single ASIC can keep the network usable.

## Backward Compatibility

Testnet 5's consensus rules are not compatible with those of Testnet 3 and Testnet 4. The
consensus rules differ in both directions: Testnet 5 enforces the BIP 54 consensus rules from

> block 1 which is not the case for Testnet 3 or Testnet 4. Testnet 5 also does not apply the

Reviewer 4: ```suggestion
block 1, which is not the case for Testnet 3 or Testnet 4. Testnet 5 also does not apply the
```

Author: Done

difficulty exception rule that Testnet 3 or Testnet 4 requires.

Because each network has its own genesis block, their UTXO sets are disjoint and transactions
can never be replayed across them. Address formats are shared, however, so the same address may
be reused on different networks.

Any implementation that intends to follow Testnet 5 must add the new network parameters and
additionally enforce the BIP 54 rules while not permitting any of the difficulty exception rules.

## Reference Implementation

> Pull request at ?

Reviewer 4: Note that a reference implementation and test vectors (if applicable) would be needed before advancing the status from Draft to Complete.

Author: Noted, will follow-up here
